File: dsa/concepts/13_graphs/java/Bfs.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;

import com.algos.topics.models.GraphAdjLst;

public class Bfs {

	public static void main(String[] args) {
		Bfs obj = new Bfs();
		List<List<Integer>> adj = new ArrayList<>();
		for (int i = 0; i < 4; i++)
			adj.add(new ArrayList<>());

		adj.get(0).add(1);
		adj.get(0).add(2);
		adj.get(1).add(2);
		adj.get(2).add(0);
		adj.get(2).add(3);
		adj.get(3).add(3);

		System.out.println(obj.bfs(adj, 0));
	}

	// Time: O(V + E)
	// Space: O(V)
	public List<Integer> bfs(List<List<Integer>> adj, int src) {
		int n = adj.size();
		List<Integer> ans = new ArrayList<>();

		Queue<Integer> q = new LinkedList<>();
		int[] vis = new int[n];

		q.add(src);
		vis[src] = 1;
		while (!q.isEmpty()) {
			int cur = q.remove();
			ans.add(cur);
			for (int nei: adj.get(cur)) {
				if (vis[nei] != 1) {
					vis[nei] = 1;
					q.add(nei);
				}
			}
		}

		return ans;
	}
}

File: dsa/concepts/13_graphs/python/bfs.py
from collections import deque

"""
Idea:
- Standard BFS traversal using a queue
- Start from source node
- Mark nodes visited to avoid revisiting
Time: O(V + E)
Space: O(V)
"""


def bfs(adj, src):
    n = len(adj)
    visited = [False] * n
    result = []

    q = deque([src])
    visited[src] = True

    while q:
        curr = q.popleft()
        result.append(curr)

        for nei in adj[curr]:
            if not visited[nei]:
                visited[nei] = True
                q.append(nei)

    return result


# -------------------- TEST --------------------
if __name__ == "__main__":
    adj = [[] for _ in range(4)]

    adj[0].extend([1, 2])
    adj[1].append(2)
    adj[2].extend([0, 3])
    adj[3].append(3)

    print(bfs(adj, 0))   # [0, 1, 2, 3]
